### PR TITLE
stage2: reify error sets

### PR DESCRIPTION
**Pull request by mitchellh**
`@Type()` for error sets.

This also fixes a bug in @typeInfo for error sets where I wasn't wrapping the payload properly. This test also catches that.

Question: I create an `error_set_merged` since that only requires a name map. Should I be creating a full `error_set` decl for this? That requires a full `Module.ErrorSet`. I'm sure I can do that but wasn't sure what you wanted here.

### CODE REVIEW

File: src/Sema.zig
Comment (andrewrk) [inline src/Sema.zig:12198]: Check out `zirOptionalPayload` for an example of how to handle the comptime optional here:

https://github.com/ziglang/zig/blob/6f303c01f3e06fe8203563065ea32537f6eff456/src/Sema.zig#L5489-L5497

File: src/Sema.zig
Comment (andrewrk) [inline src/Sema.zig:12225]: This is an internal compiler error - effectively an assertion failure - so it shouldn't use the normal compile error mechanism.

```suggestion
            else => unreachable,
```

If we wanted to make a more fail-safe build of the compiler we would probably do it by using -OReleaseSafe and having a custom panic handler that makes an error report. In such case the above `unreachable` is all it would take to trigger the appropriate mechanism.

But I think this comment becomes irrelevant after the other comment above is addressed.

**Review by andrewrk — CHANGES_REQUESTED** [review]
Nice work. I have a requested change that I think will be worth doing before merging.

> Question: I create an `error_set_merged` since that only requires a name map. Should I be creating a full `error_set` decl for this? That requires a full `Module.ErrorSet`. I'm sure I can do that but wasn't sure what you wanted here.

Just the `error_set_merged` is best for now. In the future we may want to complicate it so that we can trace compile errors back to which decls errors came from, and attached doc comments for generated documentation or perhaps IDE integration, but for now sticking with the simplest thing that works is the way to go. Easy to add these mechanisms later when we are focusing directly on them.

File: src/Sema.zig
Comment (mitchellh) [reply src/Sema.zig]: I took a look, I'm unsure exactly what you wanted me to do. See the added commit but I'm not sure this is what you intended. Sorry for the confusion, if you can clarify or reword I'll try again.

File: src/Sema.zig
Comment (andrewrk) [inline src/Sema.zig:12203]: This change takes us almost there- only one more thing left to do:

```suggestion
            const payload_val = if (union_val.val.castTag(.opt_payload)) |p| p.data else union_val.val;
            const slice_val = payload_val.castTag(.slice).?.data;
```

Optionals can be represented both ways. This should probably be abstracted into a helper method just like `isNull` so that Sema code does not rely too heavily on checking tag(). (that's not going to block merging this though)

**Review by andrewrk — CHANGES_REQUESTED** [review]
One last request and then this is good to go as soon as it passes the tests.

File: src/Sema.zig
Comment (mitchellh) [reply src/Sema.zig:12203]: Oooooooh. This explains why my "bug fix" in this PR also didn't cause issues unless it was deeply introspected using the raw values/types. Okay. Yes, I understand. I'll poke around and might do the helper too.

Comment (andrewrk) [reply src/Sema.zig:12203]: For some more context: comptime value representation upholds the property that if `coerceInMemoryAllowed` returns `ok` then it is safe to simply change the Type while keeping the Value the same. In other words `coerceInMemoryAllowed` applies to both runtime representation and comptime representation! As a consequence, optional pointer types such as `?*i32` support the same Value representation as non-optional pointers such as `*i32`.

Comment (mitchellh) [reply src/Sema.zig]: Makes sense. Thank you. The change is in 👍

File: src/Sema.zig
Comment (andrewrk) [inline src/Sema.zig:12199]:
```suggestion
            const payload_val = union_val.val.optionalValue() orelse
                return sema.addType(Type.initTag(.anyerror));
```

**Review by andrewrk — APPROVED** [review]
logic looks right! this is good to merge when tests pass.

I left one piece of style feedback but it's not important.

File: src/Sema.zig
Comment (mitchellh) [reply src/Sema.zig]: Thanks, made that change! Makes sense.

**Comment by andrewrk** [comment]
Tests passed for me locally.

**Pull request merged**
